File: app/Models/Category.php
<?php

namespace App\Models;

use Illuminate\Http\Request;
use Illuminate\Support\Str;
use Laravel\Sanctum\HasApiTokens;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Notifications\Notifiable;
use Illuminate\Database\Eloquent\Factories\HasFactory;

class Category extends Model
{
    use HasApiTokens, HasFactory, Notifiable;

    // Seller category-request lifecycle (docs/CHANGELOG_FEATURE_AUDIT.md v1.0.6/v1.0.11), tracked in
    // `approval_status` alongside the pre-existing `status` (0/1/2) convention already used for products
    // (see Admin\ProductController::update_status()) - `status` drives every existing "is this usable"
    // query (2 = pending is never selected by `where('status', 1)`), `approval_status` exists only so a
    // rejected request can stay visible/distinguishable in the seller's own request history instead of
    // looking identical to "admin deactivated an approved category".
    const APPROVAL_PENDING = 'pending';
    const APPROVAL_APPROVED = 'approved';
    const APPROVAL_REJECTED = 'rejected';

    protected $fillable = [
        'name',
        'store_id',
        'slug',
        'parent_id',
        'image',
        'banner',
        'status',
        'style',
        'row_order',
        'clicks',
        'requested_by_seller_id',
        'approval_status',
    ];

    public static function getCategories()
    {
        return static::all();
    }

    // protected static function boot()
    // {
    //     parent::boot();

    //     static::creating(function ($category) {
    //         $category->slug = Str::slug($category->name);
    //         $count = 1;
    //         while (static::whereSlug($category->slug)->exists()) {
    //             $category->slug = Str::slug($category->name) . '-' . $count++;
    //         }
    //     });
    // }

    public function products()
    {
        return $this->hasMany(Product::class, 'category_id');
    }

    public function children()
    {
        return $this->hasMany(Category::class, 'parent_id')->where('status', 1);
    }
}

File: seller_app/lib/features/profile/addProduct/models/category.dart
class Category {
  late int id;
  late int storeId;
  late String name;
  int? parentId;
  String? slug;
  late String image;
  late String banner;
  String? style;
  int? rowOrder;
  int? status;
  int? clicks;
  String? createdAt;
  String? updatedAt;
  List<Category>? children;
  String? text;
  CatState? state;
  String? icon;
  int? level;
  int? total;

  Category(
      {required id,
      required storeId,
      required name,
      parentId,
      slug,
      required image,
      required banner,
      style,
      rowOrder,
      status,
      clicks,
      createdAt,
      updatedAt,
      children,
      text,
      state,
      icon,
      level,
      total});

  Category.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    storeId = json['store_id'];
    name = json['name'];
    parentId = json['parent_id'];
    slug = json['slug'];
    image = json['image'];
    banner = json['banner'];
    style = json['style'];
    rowOrder = json['row_order'];
    status = json['status'];
    clicks = json['clicks'];
    createdAt = json['created_at'];
    updatedAt = json['updated_at'];
    children = <Category>[];
    if (json['children'] != null) {
      json['children'].forEach((v) {
        children!.add(Category.fromJson(v));
      });
    }
    text = json['text'];
    state = json['state'] != null ? CatState.fromJson(json['state']) : null;
    icon = json['icon'];
    level = json['level'];
    total = json['total'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['id'] = id;
    data['store_id'] = storeId;
    data['name'] = name;
    data['parent_id'] = parentId;
    data['slug'] = slug;
    data['image'] = image;
    data['banner'] = banner;
    data['style'] = style;
    data['row_order'] = rowOrder;
    data['status'] = status;
    data['clicks'] = clicks;
    data['created_at'] = createdAt;
    data['updated_at'] = updatedAt;
    if (children != null) {
      data['children'] = children!.map((v) => v.toJson()).toList();
    }
    data['text'] = text;
    if (state != null) {
      data['state'] = state!.toJson();
    }
    data['icon'] = icon;
    data['level'] = level;
    data['total'] = total;
    return data;
  }
}

class CatState {
  bool? opened;

  CatState({opened});

  CatState.fromJson(Map<String, dynamic> json) {
    opened = json['opened'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['opened'] = opened;
    return data;
  }
}
